### DISCUSSION TITLE
Would I be able to use just the AI deck to make a drone fly through a door by itself?

### DISCUSSION THREAD

**Discussion by cuber-1** [bitcraze/discussions#1690] [Q&A]
Would this be a possibility?

**Comment by gemenerik** [comment]
This must be possible, but it is not trivial at all and would involve designing your own or porting an existing neural network / CV algorithm to detect the door. That said; to give you some pointers you could take a look at monocular depth estimation, optical flow estimation, gate/gap detection.

**Reply by cuber-1** [reply]
Thank you. I was also wondering how I would go about commanding the drone to say, "go backwards" if the camera detects that it is close to a wall or something. I am on Windows OS, by the way.

**Reply by cuber-1** [reply]
Additionally, would a set frame rate be necessary for a project like this?

**Reply by cuber-1** [reply]
Do you recommend any specific datasets?

**Reply by ToveRumar** [reply]
Obstacle avoidance is also not a trivial functionality of the AI-deck. If you do not need the algorithm to be very sophisticated, this can be done by setting a new set point via CPX.  Please see the following documentation for pointers; 
https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/other-examples/stm_gap8_cpx/ 
https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/crtp/crtp_commander/
https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/commanders_setpoints/


I would say a fixed frame rate is always easier to work with, however a non fixed could probably also work fine for the avoidance part.

Reagarding the datasets for this, I would suggest looking into the re-search that has been done using the AI-deck, in this blog category https://www.bitcraze.io/category/ai-deck/

I hope this could point you in the right direction.

**Reply by cuber-1** [reply]
Thank you so much for all the details! How would I manage a fixed frame rate?

**Reply by cuber-1** [reply]
Also, I feel like everything I've seen research-wise has integrated the multi-ranger deck in some fashion to help with avoidance. Is there anything with just the AI Deck?

**Reply by cuber-1** [reply]
Additionally, are there specific examples available anywhere with the Commander Module being utilized?

**Reply by ToveRumar** [reply]
Hi!

At this date I do not know of any re-search on object detaction being done with only the AI-deck. I do think this might be very difficult (if possible) to do. 
I think one of the bigger issues here is distance estimation based only on one camera (monocular depth estimation). I think you need more sensory input to be able
to adequately teach the drone what to avoid, how and when. 

However together with another deck, such as multi-ranger or flow deck it is a possibility. As I said this is a quite complex endeavor and I dont think anyone in the team has done any tests or examples with this specifically. 
However maybe someone else has input on how this could be done?

It is a very interesting field and I think trying this out will be very fun!
I think the topics @gemenerik  has suggested above is a good place to start when investigating this!

**Reply by cuber-1** [reply]
Thank you for the response. Do you think I could take a project solely focused on monocular depth estimation that would do something like just tell me the distance from a wall? Would that be possible with just the AI deck?

**Reply by cuber-1** [reply]
Also what is CPX?

**Reply by cuber-1** [reply]
Would I maybe be able to implement this somehow? https://towardsdatascience.com/monocular-depth-estimation-with-depth-anything-v2-54b6775abc9f

**Reply by gemenerik** [reply]
> Also what is CPX?

Check out the [documentation for CPX](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/cpx/)


> Thank you for the response. Do you think I could take a project solely focused on monocular depth estimation that would do something like just tell me the distance from a wall? Would that be possible with just the AI deck?

This is a really cool project, but it’s not an easy one. Monocular depth estimation is usually done with powerful GPUs, so you’ll need to find or design a model that works on the AI deck’s limited hardware. That means either researching lightweight methods or even training your own small-scale neural network. If you’re ready to put in the time to learn, it’s totally possible! But if you’re looking for something more plug-and-play, this might not be the best fit.

If you're new to this, I'd recommend starting with a simpler CNN first—maybe something like image classification or basic object detection—before jumping into depth estimation. That way, you can get familiar with how models run on the AI Deck without dealing with all the extra challenges right away.

**Reply by cuber-1** [reply]
I think I'd like to jump straight in given my time constraints with this project. Do you think this would work? https://www.sciencedirect.com/science/article/pii/S0924271620301544

**Reply by ToveRumar** [reply]
Hi @cuber-1 !

Unfortunately, its very hard to say if this would work or not.
I dont think I can tell you any more than that this looks really interesting, but might be time consuming.

Maybe someone else in the community have insights. 
Good luck with the project!

**Reply by cuber-1** [reply]
"If you're new to this, I'd recommend starting with a simpler CNN first—maybe something like image classification or basic object detection—before jumping into depth estimation. That way, you can get familiar with how models run on the AI Deck without dealing with all the extra challenges right away."

How would I go about doing CNN?

**Reply by gemenerik** [reply]
Great! If you want to use a CNN, the first step is learning how to build, train, and optimize one. The AI Deck is meant for developers to experiment with their own neural networks, so it's not a plug-and-play solution. While we provide some examples and you can find related research to learn from, getting something like autonomous navigation working will require designing and training your own model.

I’d recommend starting with a deep learning course or tutorials to understand the basics of CNNs. There’s a learning curve, but if you’re motivated, it’s definitely achievable!
